Gym/PPO/main.swift

```diff
@@ -0,0 +1,138 @@
+        }
+        return unwrapped
+    }
+}
+
+// Initialize Python. This comment is a hook for internal use, do not remove.
+
+let np = Python.import("numpy")
+let gym = Python.import("gym")
+let plt = Python.import("matplotlib.pyplot")
+
+
+let env = gym.make("CartPole-v0")
+let observationSize: Int = Int(env.observation_space.shape[0])!
+let actionCount: Int = Int(env.action_space.n)!
+
+// Hyperparameters
+/// The size of the hidden layer of the 2-layer actor network and critic network. The actor network
+/// has the shape observationSize - hiddenSize - actionCount, and the critic network has the same
+/// shape but with a single output node.
+let hiddenSize: Int = 128
+/// The learning rate for both the actor and the critic.
+let learningRate: Float = 0.0003
+/// The discount factor. This measures how much to "discount" the future rewards
+/// that the agent will receive. The discount factor must be from 0 to 1
+/// (inclusive). Discount factor of 0 means that the agent only considers the
+/// immediate reward and disregards all future rewards. Discount factor of 1
+/// means that the agent values all rewards equally, no matter how distant
+/// in the future they may be. Denoted gamma in the PPO paper.
+let discount: Float = 0.99
+/// Number of epochs to run minibatch updates once enough trajectory segments are collected. Denoted
+/// K in the PPO paper.
+let epochs: Int = 10
+/// Parameter to clip the probability ratio. The ratio is clipped to [1-clipEpsilon, 1+clipEpsilon].
+/// Denoted epsilon in the PPO paper.
+let clipEpsilon: Float = 0.1
+/// Coefficient for the entropy bonus added to the objective. Denoted c_2 in the PPO paper.
+let entropyCoefficient: Float = 0.0001
+/// Maximum number of episodes to train the agent. The training is terminated
+/// early if maximum score is achieved consecutively 10 times.
+let maxEpisodes: Int = 1000
+/// Maximum timestep per episode.
+let maxTimesteps: Int = 200
+/// The length of the trajectory segment. Denoted T in the PPO paper.
+let updateTimestep: Int = 1000
+
+var memory: PPOMemory = PPOMemory()
+var agent: PPOAgent = PPOAgent(
+    observationSize: observationSize,
+    hiddenSize: hiddenSize,
+    actionCount: actionCount,
+    learningRate: learningRate,
+    discount: discount,
+    epochs: epochs,
+    clipEpsilon: clipEpsilon,
+    entropyCoefficient: entropyCoefficient
+)
+
+// Training loop
+var timestep: Int = 0
+var episodeReturn: Float = 0
+var episodeReturns: [Float] = []
+var maxEpisodeReturn: Float = -1
+for episodeIndex in 1..<maxEpisodes+1 {
+    var state = env.reset()
+    for _ in 0..<maxTimesteps {
+        timestep += 1
+        let tfState: Tensor<Float> = Tensor<Float>(numpy: np.array(state, dtype: np.float32))!
+        let dist: Categorical<Int32> = agent.oldActorCritic(tfState)
+        let action: Int32 = dist.sample().scalarized()
+        let logProb: Float = Float(dist.logProbabilities.makeNumpyArray()[action])!
```

Comment:
You can replace `Tensor -> NumPy -> Float` conversion with `Tensor` indexing and `Tensor.scalarized()`.

```suggestion
        let logProb: Float = dist.logProbabilities[action].scalarized()
```